drivers/flash/soc_flash_b91.c

```diff
@@ -0,0 +1,183 @@
+	struct k_sem write_lock;
+};
+
+/* driver parameters structure */
+static const struct flash_parameters flash_b91_parameters = {
+	.write_block_size = DT_PROP(DT_INST(0, soc_nv_flash), write_block_size),
+	.erase_value = 0xff,
+};
+
+
+/* Check for correct offset and length */
+static bool flash_b91_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len < 1)) {
+		return false;
+	}
+
+	/* check for max value */
+	if ((offset + len) > FLASH_SIZE) {
+		return false;
+	}
+
+	return true;
+}
+
+/* API implementation: driver initialization */
+static int flash_b91_init(const struct device *dev)
+{
+	struct flash_b91_data *dev_data = dev->data;
+
+	k_sem_init(&dev_data->write_lock, 1, 1);
+
+	return 0;
+}
+
+/* API implementation: erase */
+static int flash_b91_erase(const struct device *dev, off_t offset, size_t len)
+{
+	int page_nums = len / PAGE_SIZE;
+	struct flash_b91_data *dev_data = dev->data;
+
+	/* return SUCCESS if len equals 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* check for valid range */
+	if (!flash_b91_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* Erase can be done only by pages */
+	if (((offset % PAGE_SIZE) != 0) || ((len % PAGE_SIZE) != 0)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	/* erase flash page by page */
+	for (int i = 0; i < page_nums; i++) {
+		flash_erase_page(offset);
+		offset += PAGE_SIZE;
+	}
+
+	/* release semaphore */
+	k_sem_give(&dev_data->write_lock);
+
+	return 0;
+}
+
+/* API implementation: write */
+static int flash_b91_write(const struct device *dev, off_t offset,
+			   const void *data, size_t len)
+{
+	struct flash_b91_data *dev_data = dev->data;
+
+	/* return SUCCESS if len equals 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* check for valid range */
+	if (!flash_b91_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	/* write flash */
+	flash_write_page(offset, len, (unsigned char *)data);
+
+	/* release semaphore */
+	k_sem_give(&dev_data->write_lock);
+
+	return 0;
+}
+
+/* API implementation: read */
+static int flash_b91_read(const struct device *dev, off_t offset,
+			  void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	/* return SUCCESS if len equals 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* check for valid range */
+	if (!flash_b91_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* read flash */
+	flash_read_page(offset, len, (unsigned char *)data);
+
+	return 0;
+}
+
+/* API implementation: get_parameters */
+static const struct flash_parameters *
+flash_b91_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return &flash_b91_parameters;
+}
+
+/* API implementation: page_layout */
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
```

Comment:
This seems redundant too.

Did you mean to use `imply FLASH_PAGE_LAYOUT` in the driver Kconfig, instead of `select`?

https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#the-imply-statement

Reply:
KConfig selects FLASH_HAS_PAGE_LAYOUT but not FLASH_PAGE_LAYOUT.
The same is implemented for all the rest Flash drivers, don't feel really safe to fix it.
Please confirm.

Reply:
Sorry, you're right. My mistake.